src/preact/component/types.d.ts

```diff
@@ -0,0 +1,59 @@
+import * as Preact from 'preact';
+import {ReadyState_Enum} from '#core/constants/ready-state';
+import {Ref, VNode, FunctionComponent, ComponentChildren} from 'preact';
+
+// TODO: figure out the proper type for all of the 'as' types.
+
+export interface WrapperComponentProps {
+  style: Preact.JSX.CSSProperties;
+  class: string;
+  as: FunctionComponent | any;
+  wrapperClassName?: string | null;
+  wrapperStyle?: Object | null;
+  children?: ComponentChildren;
+}
+
+/**
+ * See https://developer.mozilla.org/en-US/docs/Web/CSS/contain
+ */
+export interface ContainerWrapperComponentProps {
+  as?: FunctionComponent | any;
+  contentAs?: string | FunctionComponent;
+  contentProps?: any;
+  style?: Preact.JSX.CSSProperties;
+  size?: boolean;
+  layout?: boolean;
+  paint?: boolean;
+  wrapperClassName?: string | null;
+  wrapperStyle?: Object | null;
+  contentRef?: null | Ref<any>;
+  contentClassName?: string | null;
+  contentStyle?: Object | null;
+  children?: ComponentChildren | null;
+  class?: string | null;
+}
+
+export type RendererFunction = (obj: JsonObject) => VNode | Promise<VNode>;
```

Comment:
How is this parsed? ie
`((obj: JsonObject) => VNode) | Promise<VNode>`
vs
`(obj: JsonObject) => (VNode | Promise<VNode>)`

Recommend adding parentheses (if linter allows) to disambiguate

Reply:
Eh, I think this is something you'll pick up. It is the latter due to operator precedence. The former would need extra parens